support qnn mean (dim=None)

cccclai opened the pull request:
Summary: Address mean op lower failure. When dim is not specified, it will take mean across all axes. For QNN, we need to get axes based on input shape

Differential Revision: D83520776

haowhsu-quic commented on backends/qualcomm/tests/test_qnn_delegate.py:
```diff
@@ -1025,6 +1025,13 @@ def test_qnn_backend_mean_dim(self):
                 self.lower_module_and_test_output(module, sample_input)
 
+    def test_qnn_backend_mean(self):
+        modules = [Mean(), Mean()]  # noqa: F405
```
Do you need more configurations here? like `Mean(dim=0)`, `Mean(dim=0, keepdim=True)`.

cccclai replied:
updated

cccclai commented on backends/qualcomm/tests/models.py:
```diff
@@ -1348,6 +1348,22 @@ def forward(self, x):
         return torch.mean(x, (-1, -2))
 
 
+class Mean(torch.nn.Module):
```
Yes updated

cccclai commented:
Does it look good? I think it can fix 5 failing op tests

haowhsu-quic commented on backends/qualcomm/tests/test_qnn_delegate.py:
```diff
@@ -2666,13 +2710,59 @@ def test_qnn_backend_max_pool2d(self):
+                module = self.get_qdq_module(
+                    test[QCOM_MODULE], test[QCOM_SAMPLE_INPUTS]
+                )
+                module = self.get_qdq_module(module, test[QCOM_SAMPLE_INPUTS])
```
Might need to remove L2759~2762, thanks.

cccclai replied:
Good catch, thank you for the review

chenweng-quic commented on backends/qualcomm/builders/op_mean_dim.py:
```diff
@@ -40,7 +40,16 @@ def define_node(
+        rank = len(input_node.meta["val"].shape)
+        dim_arg = node.args[1]
+
+        if dim_arg is None:
```
Maybe edge case, could dim_arg be empty list?

cccclai replied:
Yes I verify the eager mode and it does support empty list. Updated

cccclai commented on backends/qualcomm/tests/test_qnn_delegate.py:
```diff
@@ -1018,12 +1018,53 @@ def test_qnn_backend_max_pool2d(self):
+            # Scalar case
+            {
+                QCOM_MODULE: Mean(),
+                QCOM_SAMPLE_INPUTS: (torch.tensor([5.0]),),
```
Thanks, good catch

cccclai replied:
Actually this is indeed a good test. ReduceMean doesn't support 0d tensor, I think we need to have a pass to convert 0d to 1d tensor if the input of mean is 0d. I have some sketch code here. What do you think? For now, I comment out the test case and think we can follow up on this
```
import torch
from executorch.exir.pass_base import ExportPass, PassResult

class Rank0ToRank1(ExportPass):
    """
    For selected ops and selected input positions, if the input is rank-0 (scalar),
    insert a reshape to [1] before the op.
    """

    def __init__(self, op_input_map=None) -> None:
        super().__init__()
        # key is the op, value is the input indices to be reshaped
        self.op_input_map = {
          torch.ops.aten.mean.dim: [0],
        }

    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
        graph = graph_module.graph
        changed = False

        for node in list(graph.nodes):
            if node.op == "call_function" and node.target in self.op_input_map:
                input_indices = self.op_input_map[node.target]

                new_args = list(node.args)
                for idx in input_indices:
                    if idx < len(new_args):
                        inp = new_args[idx]
                        if hasattr(inp, "meta"):
                            val = inp.meta.get("val", None)
                            if val is not None and hasattr(val, "shape") and val.shape == ():
                                # Insert reshape right before the op
                                with graph.inserting_before(node):
                                    reshape_node = graph.call_function(
                                        torch.ops.aten.reshape.default,
                                        args=(inp, (1,))
                                    )
                                    reshape_node.meta["val"] = val.reshape(1,)

                                # Replace arg idx with reshape_node
                                new_args[idx] = reshape_node
                                changed = True

                # update node args if modified
                node.args = tuple(new_args)

        if changed:
            graph_module.recompile()
        return PassResult(graph_module, changed)

```

cccclai commented on backends/qualcomm/tests/test_qnn_delegate.py:
```diff
@@ -2666,13 +2707,55 @@ def test_qnn_backend_max_pool2d(self):
             with self.subTest(i=i):
-                module = self.get_qdq_module(module, sample_input)
-                self.lower_module_and_test_output(module, sample_input)
+                module = self.get_qdq_module(module, test[QCOM_SAMPLE_INPUTS])
```
Sorry I should be more careful..

cccclai committed:
support qnn mean (dim=None) (#14675)

Summary:

Address mean op lower failure. When dim is not specified, it will take mean across all axes. For QNN, we need to get axes based on input shape

Differential Revision: D83520776

cccclai commented:
@pytorchbot cherry-pick --onto release/1.0 -c regression
